esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,532 @@ int CEspHttpServer::onGet()
+        {
+            const char* loginURL = authReq.authBinding->queryLoginURL();
+            if (!strieq(loginURL, redirectURL))
+                m_response->redirect(*m_request, redirectURL);
+            else
+                m_response->redirect(*m_request, "/");
+        }
+        else
+            m_response->redirect(*m_request, "/");
+    }
+    else if (authReq.methodName && strieq(authReq.methodName, "logout"))
+    {
+        clearCookie(authReq.authBinding->querySessionIDCookieName());
+        const char* logoutURL = authReq.authBinding->queryLogoutURL();
+        if (!isEmptyString(logoutURL))
+            m_response->redirect(*m_request, authReq.authBinding->queryLogoutURL());
+        else
+            send200OK();
+    }
+}
+
+void CEspHttpServer::handleAuthFailed(bool sessionAuth, EspAuthRequest& authReq, unsigned* sessionID)
+{
+    ISecUser *user = authReq.ctx->queryUser();
+    if (user && (user->getAuthenticateStatus() == AS_PASSWORD_EXPIRED || user->getAuthenticateStatus() == AS_PASSWORD_VALID_BUT_EXPIRED))
+    {
+        ESPLOG(LogMin, "ESP password expired for %s", authReq.ctx->queryUserId());
+        handlePasswordExpired(sessionAuth);
+        return;
+    }
+
+    if (!sessionAuth)
+    {
+        ESPLOG(LogMin, "Authentication failed: send BasicAuthentication.");
+        m_response->sendBasicChallenge(authReq.authBinding->getChallengeRealm(), true);
+        return;
+    }
+
+    ESPLOG(LogMin, "Authentication failed: call askUserLogin.");
+    if (sessionID)
+        askUserLogin(authReq.authBinding, *sessionID);
+    else
+        askUserLogin(authReq.authBinding, createHTTPSession(authReq, nullptr));
+}
+
+void CEspHttpServer::askUserLogin(EspHttpBinding* authBinding, unsigned sessionID)
+{
+    VStringBuffer sessionIDStr("%u", sessionID);
+    addCookie(authBinding->querySessionIDCookieName(), sessionIDStr.str(), authBinding->getSessionTimeoutSeconds());
+    addCookie(SESSION_AUTH_COOKIE, "1", 0); //time out when browser is closed
+    m_response->redirect(*m_request, authBinding->queryLoginURL());
+}
+
+void CEspHttpServer::send200OK()
+{
+    m_response->setContentType("text/html; charset=UTF-8");
+    m_response->setStatus(HTTP_STATUS_OK);
+    m_response->send();
+}
+
+unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, const char* loginURL)
+{
+    Owned<IRemoteConnection> conn = querySDSConnection(authReq.authBinding->queryDomainSessionSDSPath(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
+    return createHTTPSession(authReq, conn->queryRoot(), loginURL);
+}
+
+unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, IPropertyTree* domainSessions, const char* _loginURL)
+{
+    CDateTime now;
+    now.setNow();
+    time_t createTime = now.getSimple();
+
+    StringBuffer peer;
+    VStringBuffer idStr("%s_%ld", m_request->getPeer(peer).str(), createTime);
+    unsigned sessionID = hashc((unsigned char *)idStr.str(), idStr.length(), 0);
```

Comment:
Yes, someone may login from both ECLWatch and WsECL which should be different sessions.